pkg/controller/deployment/sync.go

```diff
@@ -294,6 +294,7 @@ func (dc *DeploymentController) getNewReplicaSet(deployment *extensions.Deployme
 			Name:            deployment.Name + "-" + podTemplateSpecHash,
 			Namespace:       namespace,
 			OwnerReferences: []metav1.OwnerReference{*newControllerRef(deployment)},
+			Finalizers:      []string{metav1.FinalizerDeleteDependents},
```

Comment:
I don't think this can work unless https://github.com/kubernetes/kubernetes/pull/46500 is merged but I am willing to be proven wrong